views/js/runner/plugins/controls/progressbar/plugin.js

```diff
@@ -0,0 +1,154 @@
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2016-2018 (original work) Open Assessment Technologies SA ;
+ */
+
+/**
+ * Test Runner Control Plugin : Progress Bar factory
+ *
+ * @author Bertrand Chevrier <bertrand@taotesting.com>
+ * @author Jean-Sébastien Conan <jean-sebastien@taotesting.com>
+ */
+define([
+    'lodash',
+    'taoTests/runner/plugin',
+    'taoQtiTest/runner/helpers/map',
+    'taoQtiTest/runner/helpers/progress',
+    'taoQtiTest/runner/plugins/controls/progressbar/renderer/percentage',
+    'taoQtiTest/runner/plugins/controls/progressbar/renderer/position'
+], function (_, pluginFactory, mapHelper, progressHelper, percentageRendererFactory, positionRendererFactory) {
+    'use strict';
+
+    /**
+     * List of default available progress indicator renderers
+     * @type {Object}
+     */
+    var defaultRenderers = {
+        percentage: percentageRendererFactory,
+        position: positionRendererFactory
+    };
+
+    /**
+     * Simple configurator that leaves the config unchanged
+     * @param {Object} config
+     * @returns {Object}
+     */
+    function defaultConfigurator(config) {
+        return config;
+    }
+
+    /**
+     * Returns the configured plugin
+     * @param {String} pluginName - The name of the plugin
+     * @param {Function} [configurator] - A callback that will customize the config
+     * @param {Object} [renderers] - A list of renderers to assign
+     * @returns {plugin} Returns a tweaked progressbar plugin
+     */
+    return function progressbarPluginFactory(pluginName, configurator, renderers) {
```

Comment:
I think this pattern is the inverse of what it should be.
I think the plugin itself shouldn't be wrapped in a factory. The plugin remains the same, it react to the same events, etc. 
However during the `install` or the `init` step the plugin selects the progress component it should use based on the configuration.

My concern is that adds a new step in the plugin loading mecanism. Here you want to configure another rendering of the plugin, but this can be anything else, so the feature configuration can be just setting a configuration option.

Reply:
The plugin already uses configuration options, the way you describe it.
The wrapper/factory adds the ability to package the plugin. The goal is to bring the ability to package plugin with forced options. This is to be used along with the `feature set` that can be chosen from the delivery (check boxes). In that context we may only select which plugin to enable/disable, not the options we want.